source/daplink/cmsis-dap/DAP.c

```diff
@@ -1791,5 +1789,22 @@ void DAP_Setup(void) {
   DAP_Data.jtag_dev.count = 0U;
 #endif
 
+  if (DAP_DEFAULT_SWJ_CLOCK >= MAX_SWJ_CLOCK(DELAY_FAST_CYCLES)) {
+    DAP_Data.fast_clock  = 1U;
+    DAP_Data.clock_delay = 1U;
+  } else {
+    DAP_Data.fast_clock  = 0U;
+
+    uint32_t delay = ((CPU_CLOCK/2U) + (DAP_DEFAULT_SWJ_CLOCK - 1U)) / DAP_DEFAULT_SWJ_CLOCK;
+    if (delay > IO_PORT_WRITE_CYCLES) {
+      delay -= IO_PORT_WRITE_CYCLES;
+      delay  = (delay + (DELAY_SLOW_CYCLES - 1U)) / DELAY_SLOW_CYCLES;
+    } else {
+      delay  = 1U;
+    }
+
+  DAP_Data.clock_delay = delay;
+  }
+
```

Comment:
A bit more info on this change from @gerargz:


> #### HIC Specific
> Changed the way the KL27 accesses the GPIO for SWCLK and SWDIO. Using the IOPORT interface to access the GPIO (FGPIO address) makes this a single cycle access instead of the three cycles it takes when accessing the GPIO through the peripheral bridge. This change now matches with IO_PORT_WRITE_CYCLES definition.
> 
> Changed the default SWCLK speed for the KL27 HIC from 5MHz to 8MHz (nrf max).
> 
> #### DAPLink core
> Changed the default way that DAPLink calculates the initial clock delay to achieve the desired SWCLK speed.
> 
> #### Board specific
> Added macro DELAY_FAST_CYCLES=2U to the microbit yaml file to allow 8MHz calculation.
> 
> ### Background
> DAPLink inserts delays to approximate the SWCLK speed to the one configured. There are two methods for inserting these delays, the “fast” delay and “slow” delay. The fast delay uses a fixed number of precompiled “nop” instructions (up to 3) and the slow delay uses a loop to have a wider range of clock delays but this has a 3 cycle granularity per loop.
> 
> In DAP_Setup() the clock_delay is calculated without considering that the “slow” delay takes 3 cycles for each iteration, so the resulting clock speed will have extra delay. In contrast, when using a DAP command (DAP_SWJ_Clock()) to change the SWCLK speed, the clock_delay does consider the extra cycles for the “slow” delay and is calculated correctly. In this PR, the method of calculating the clock delay from DAP_SWJ_Clock() was added to DAP_Setup().
> 
> To generate an 8MHz SWCLK for the nrf, the “fast” delay needs to be used with a 2 cycle delay. With a 48MHz KL27 core clock, one cycle for the FGPIO access and 2 cycle fast delay, this results in a 8MHz SWCLK speed ((48MHz/2)/(1+2)=8MHz). The macro DELAY_FAST_CYCLES=2U was added to the microbit yaml file.
>

Reply:
This is a file we keep in sync with CMSIS-DAP.

- Will this change be submitted to CMSIS-DAP?
- If we are to be keeping this change compared to upstream for some time, I would rather keep the new lines where the old ones are removed.